src/main/java/fi/aalto/cs/apluscourses/presentation/AuthenticationViewModel.java

```diff
@@ -13,13 +15,24 @@
   @NotNull
   private final String authenticationUrl;
 
+  @NotNull
+  private final ExerciseDataSource exerciseDataSource;
+
+  @Nullable
+  private Authentication authentication;
```

Comment:
Have you considered synchronization of this mutable field? Is there possibility that multiple threads access it?

Reply:
I learned about synchronization after making this pr, so I haven't considered it yet